sklearn/decomposition/tests/test_pca.py

```diff
@@ -15,37 +18,118 @@
+
+        # Test get_covariance and get_precision
+        cov = pca.get_covariance()
+        precision = pca.get_precision()
+        assert_array_almost_equal(np.dot(cov, precision),
+                                  np.eye(X.shape[1]), 12)
+
+    # test explained_variance_ratio_ == 1 with all components
+    pca = PCA(svd_solver='full')
     pca.fit(X)
     assert_almost_equal(pca.explained_variance_ratio_.sum(), 1.0, 3)
 
-    X_r = pca.transform(X)
-    X_r2 = pca.fit_transform(X)
 
-    assert_array_almost_equal(X_r, X_r2)
+def test_pca_arpack_solver():
+    # PCA on dense arrays
+    X = iris.data
+    d = X.shape[1]
+
+    # Loop excluding the extremes, invalid inputs for arpack
+    for n_comp in np.arange(1, d):
+        pca = PCA(n_components=n_comp, svd_solver='arpack', random_state=0)
+
+        X_r = pca.fit(X).transform(X)
+        np.testing.assert_equal(X_r.shape[1], n_comp)
 
-    # Test get_covariance and get_precision with n_components == n_features
-    # with n_components < n_features and with n_components == 0
-    for n_components in [0, 2, X.shape[1]]:
-        pca.n_components = n_components
-        pca.fit(X)
+        X_r2 = pca.fit_transform(X)
+        assert_array_almost_equal(X_r, X_r2)
+
+        X_r = pca.transform(X)
+        X_r2 = pca.fit_transform(X)
+        assert_array_almost_equal(X_r, X_r2)
+
+        # Test get_covariance and get_precision
+        cov = pca.get_covariance()
+        precision = pca.get_precision()
+        assert_array_almost_equal(np.dot(cov, precision),
+                                  np.eye(d), 12)
+
+    pca = PCA(n_components=0, svd_solver='arpack', random_state=0)
+    assert_raises(ValueError, pca.fit, X)
+    assert_equal(pca.n_components,
+                 PCA(n_components=0,
+                     svd_solver='arpack', random_state=0).n_components)
+    assert_equal(pca.svd_solver,
+                 PCA(n_components=0,
+                     svd_solver='arpack', random_state=0).svd_solver)
+
+    pca = PCA(n_components=d, svd_solver='arpack', random_state=0)
+    assert_raises(ValueError, pca.fit, X)
+    assert_equal(pca.n_components,
+                 PCA(n_components=d,
+                     svd_solver='arpack', random_state=0).n_components)
+    assert_equal(pca.svd_solver,
+                 PCA(n_components=0,
+                     svd_solver='arpack', random_state=0).svd_solver)
+
+
+def test_pca_randomized_solver():
+    # PCA on dense arrays
+    X = iris.data
+
+    # Loop excluding the extremes, invalid inputs for arpack
+    for n_comp in np.arange(1, X.shape[1]):
+        pca = PCA(n_components=n_comp, svd_solver='randomized', random_state=0)
+
+        X_r = pca.fit(X).transform(X)
+        np.testing.assert_equal(X_r.shape[1], n_comp)
+
+        X_r2 = pca.fit_transform(X)
+        assert_array_almost_equal(X_r, X_r2)
+
+        X_r = pca.transform(X)
+        X_r2 = pca.fit_transform(X)
+        assert_array_almost_equal(X_r, X_r2)
```

Comment:
Ah, so you're implicitly testing that calling ``transform`` doesn't change the ``fit`` parameters? If that's so, maybe that component of the test should be separate.